### PR TITLE
 Thumbnails, Characterization, and Virus-Checking (Oh my!)

### LINKED ISSUE #19: Derivatives Generation Setup
Need to actually set up our derivatives infrastructure

-  Add sidekiq to project
- Spin off derivative job after item creation

### LINKED ISSUE #402: maintain file name when downloading
The file deposited through UI doesn't seem to use its' original filename when downloading? it uses a uuid instead:
http://uat.library.ualberta.ca/items/6c6397b9-f292-4542-b418-6f1db50bd4ef/download/2r36tz89220180126-21-n5cso6

### PR DESCRIPTION
**Pull request by mbarnett**
Closes #19 

Also closes #402 and #362.

I've made virus-checking and FITS optional for development purposes. For UAT, Staging, and Production, clamav should be running as a daemon on the machine, and fits needs to be installed and in the Rails' process' user's path.

@weiweishi For your migration script, for each item or thesis, you'll need to call `thumbnail_fileset` on the item after saving, and pass it a fileset on that item that contains the file that should be used to derive the thumbnail. Presumably this will be the first (only) file on most items, since we didn't support multi-file items in Hydranorth, so you probably just need to do something like:

```ruby
item.thumbnail_fileset(item.file_sets.first) 
```

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -89,3 +89,7 @@ group :test do
   gem 'minitest-hooks'
   gem 'shoulda', require: false
 end
+
+group :uat, :staging, :production do
+  gem 'clamby'
```
Comment (murny) [inline Gemfile]: Is there any instructions to install this? Would that be worth adding to readme or something? Or maybe just link to its readme: https://github.com/kobaltz/clamby#dependencies

File: app/models/concerns/item_properties.rb [hunk 2]
```diff
@@ -169,8 +172,22 @@ def each_community_collection
...
-    nil
+    @thumbnail ||= ActiveStorage::Attached::One.new(:thumbnail, self)
+  end
+
+  def thumbnail_fileset(fileset)
+    raise ArgumentError, 'Thumbnail must belong to the item it is set for' unless id == fileset.owning_item.id
+    thumbnail.purge if thumbnail.present?
+    fileset.unlock_and_fetch_ldp_object do |unlocked_fileset|
+      unlocked_fileset.create_derivatives
+      # don't ask. RDF::URIs aren't real Ruby URIs for reasons that presumably made sense to someone, somewhere
```
Comment (murny) [inline app/models/concerns/item_properties.rb]: :upside_down_face:

File: app/models/draft_item.rb [hunk 3]
```diff
@@ -185,30 +197,34 @@ def validate_if_visibility_is_embargo?
...
+    files.each do |file|
+      path = file_path_for(file)
+      errors.add(:files, :infected, filename: file.filename.to_s) unless Clamby.safe?(path)
+    end
+  end
+
   # Fedora file handling
   # Convert ActiveStorage objects into File objects so we can deposit them into fedora
   #
   # TODO: How to handle thumbnail? Should it be the first file uploaded to fedora?
```
Comment (murny) [inline app/models/draft_item.rb]: Can probably delete this TODO now since its now fixed!

File: app/models/concerns/item_properties.rb (hunk 2 above)
Comment (mbarnett) [reply app/models/concerns/item_properties.rb]: I don't know man

File: Gemfile (hunk 1 above)
Comment (mbarnett) [reply Gemfile]: I'll add a link to the Readme for this and FITS

**Comment by mbarnett** [comment]
Added some more comments, so you'll need to re-approve

**Pull request merged**